router/src/main/java/io/micronaut/web/router/version/RoutesVersioningConfiguration.java

```diff
@@ -0,0 +1,165 @@
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.core.util.Toggleable;
+
+import static io.micronaut.web.router.version.RoutesVersioningConfiguration.PREFIX;
+
+/**
+ * Routes versioning configuration.
+ *
+ * @author BogdanOros
+ */
+@ConfigurationProperties(PREFIX)
+@Requires(property = PREFIX + ".enabled", value = StringUtils.TRUE)
+public class RoutesVersioningConfiguration implements Toggleable {
+
+    /**
+     * The configuration property.
+     */
+    public static final String PREFIX = "micronaut.router.versioning";
+
+    /**
+     * The default enable value.
+     */
+    private static final boolean DEFAULT_ENABLED = false;
+
+    /**
+     * The enable value.
+     */
+    private boolean enabled = DEFAULT_ENABLED;
+
+    /**
+     * Request header based versioning configuration.
+     */
+    private HeaderBasedVersioningConfiguration header = new HeaderBasedVersioningConfiguration();
+
+    /**
+     * Request parameter based versioning configuration.
+     */
+    private ParameterBasedVersioningConfiguration parameter = new ParameterBasedVersioningConfiguration();
+
+    /**
+     * @return The header based versioning configuration.
+     */
+    public HeaderBasedVersioningConfiguration getHeader() {
+        return header;
+    }
+
+    /**
+     * @param header The header based versioning configuration.
+     */
+    public void setHeader(HeaderBasedVersioningConfiguration header) {
+        this.header = header;
+    }
+
+    /**
+     * @return The request parameter based versioning configuration.
+     */
+    public ParameterBasedVersioningConfiguration getParameter() {
+        return parameter;
+    }
+
+    /**
+     * @param parameter The request parameter based versioning configuration.
+     */
+    public void setParameter(ParameterBasedVersioningConfiguration parameter) {
+        this.parameter = parameter;
+    }
+
+    /**
+     * @param enabled Enables the version based route matches filtering.
+     */
+    public void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+    }
+
+    /**
+     * @return {@code true} if version based matches filtering is enabled.
+     */
+    @Override
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    /**
+     * Abstract specific configuration for different versioning configurations.
+     */
+    private abstract static class AbstractVersioningConfiguration implements Toggleable {
+
+        private boolean enabled;
+        private String name;
+
+        AbstractVersioningConfiguration(String name) {
+            this.name = name;
+        }
+
+        /**
+         * @return the property name to extract specified version from.
+         */
+        public String getName() {
+            return name;
+        }
+
+        /**
+         * @param name The property name to extract specified version from.
+         */
+        public void setName(String name) {
```

Comment:
Maybe we should allow multiple header / parameter names? Some APIs allow for example `Accept-Version`

Reply:
You mean different header keys? For example: "VERSION, X-VERSION, API-VERSION" (listed in properties) and the extractor will look for a match?

Reply:
Yes